tests/test_hoist.py

```diff
@@ -0,0 +1,264 @@
+
+    # compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+    # Analysis
+    scheduler.process(transformation=HoistAnalysisFindVariables(), reverse=True)
+    # Synthesis
+    scheduler.process(transformation=HoistSynthesis())
+
+    for item in scheduler.items:
+        if write_transformation_output:
+            suffix = '.find_variables.out.F90'
+            sourcefile = item.source
+            sourcefile.write(path=sourcefile.path.with_suffix(suffix).name)
+        if "driver" in item.name:
+            assert len(item.routine.arguments) == 3
+        elif "kernel1" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "kernel2" in item.name:
+            assert len(item.routine.arguments) == 5
+        elif "device1" in item.name:
+            assert len(item.routine.arguments) == 5
+        elif "device2" in item.name:
+            assert len(item.routine.arguments) == 4
+
+    compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_hoist_allocatable(here, frontend, config):
+    """
+    Testing hoist functionality with modified *Analysis* part using a `find_variables` method only searching for arrays
+    with variable `a` within the shape/dimensions **and** modified *Synthesis* part using `driver_variable_declaration`
+    implementation declaring hoisted arrays as *allocatable*, including allocation and de-allocation.
+    """
+
+    class HoistAnalysisFindVariables(HoistAnalysis):
+
+        @staticmethod
+        def find_variables(routine):
+            return [var for var in routine.variables if var not in routine.arguments and isinstance(var, sym.Array)
+                    and "a" in FindVariables().visit(var.dimensions)]
+
+    class HoistSynthesisAllocatable(HoistSynthesis):
+
+        @staticmethod
+        def driver_variable_declaration(routine, var):
+            dimensions = [sym.IntrinsicLiteral(":")] * (len(var.dimensions))
+            routine.spec.append(ir.VariableDeclaration((var.clone(dimensions=as_tuple(dimensions),
+                                                                  type=var.type.clone(allocatable=True)),)))
+            routine.body.prepend(ir.Allocation((var.clone(),)))
+            routine.body.append(ir.Deallocation((var.clone(dimensions=None),)))
+
+    proj = here/'sources/projHoist'
+
+    config['routine'] = [
+        {
+            'name': 'driver',
+            'role': 'driver',
+            'expand': True,
+        },
+    ]
+
+    scheduler = Scheduler(paths=[proj], config=config, seed_routines=['driver'], frontend=frontend)
+
+    # compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+    # Analysis
+    scheduler.process(transformation=HoistAnalysisFindVariables(), reverse=True)
+    # Synthesis
+    scheduler.process(transformation=HoistSynthesisAllocatable())
+
+    for item in scheduler.items:
```

Comment:
Can you add here some tests that verify the attributes of the hoisted variables and the other code changes? For example, that the allocation nodes exist?

Otherwise it might just have been working by accident and not really testing the actual intrinsics of that transformation mode.